### PR TITLE
using dedicated HTTP clients

### PR DESCRIPTION
**Pull request by arschles**
This attempts to use a dedicated, global HTTP client with a configurable global timeout for all scalers that do HTTP operations. This is a draft PR for now, but I'll update the docs too if the direction is acceptable.

### Checklist

- [x] Commits are signed with Developer Certificate of Origin (DCO)
- [x] Tests have been added
- [x] A PR is opened to update the documentation on https://github.com/kedacore/keda-docs (see https://github.com/kedacore/keda-docs/pull/324)
- [x] Changelog has been updated
- [x] Helm charts have been updated

Relates to https://github.com/kedacore/keda/issues/1133

Much of this was done with the help of @iennae

### CODE REVIEW

**Comment by zroubalik** [comment]
@aman-bansal PTAL^

**Comment by aman-bansal** [comment]
Sure I will take a look by EOD

File: pkg/scalers/azure_blob_scaler.go
Comment (aman-bansal) [inline pkg/scalers/azure_blob_scaler.go:47]: Changing the contract for creating a new scaler seems unnecessary. I very much like the recent changes where we simplified the creation of a scaler object with ScalerConfig. It's a cleaner approach.

File: adapter/main.go
Comment (aman-bansal) [inline adapter/main.go:112]: I don't think http-timeout is that of a critical parameter that should be part of the application start. I guess having something of say 3 sec as constant is fine. If a client wants to specify a different timeout he can simply add it to trigger metadata of config. It seems neat to me.

@zroubalik @turbaszek do you any thoughts in this regard?

Comment (zroubalik) [reply adapter/main.go:112]: I am not 100% against this, but on the other hand, I can see your point. I honestly don't have a strict opinion on this 🤷‍♂️ I am open to hear others ideas.

Comment (arschles) [reply adapter/main.go:112]: My thought process was that this would give a single timeout for all triggers. Would you rather go with a per-trigger timeout?

Comment (aman-bansal) [reply adapter/main.go:112]: okay here are the possible solutions I see. I might be missing few things so do let me know if I am wrong anywhere.

How I view keda is something like this, we start keda with few flags that manages the state of keda operators and keda metric server. Each scaler manages its own state based on trigger metadata(including auth meta). I don't see any information that passes from keda to scalers as such. ScalerConfig is the example of that. Scaler config is just basic name and metadata information. Keda manages which scaler to run, after that each scaler is independent. 

Now with this issue what we are trying is pass some control parameters to each scaler, and that's where the problem is we don't have any implementation regarding this (atleast i don't see anything).

So I see three different solutions for this:
1. Introduce global variable. Though I don't see any global variables in the code right now. This would introduce complexity while testing and all the know drawbacks of global variables.

2. Expand Scaler Config to have some parameters of Keda. This will help to extend more control parameters that might needed in future. But this is increasing the scope. I don't think this issue needs that much of rework. But yes something to think of and maybe build use cases around this.

3. Introduce a new env variable. We will pick it from the env variable. This is easy, neat and std industry practice. We could have something like this KEDA_HTTP_DEFAULT_TIMEOUT. This will do the trick.

@zroubalik @arschles What do you think?

Comment (arschles) [reply adapter/main.go:112]: Apologies for the delay, I had to give a talk yesterday.

@aman-bansal I agree regarding 1, so I would vote against that personally.

I think I'm on the same page as you regarding 2 and 3. Assuming we have a global timeout variable that you pass into KEDA when you start it up, you could also allow people to override that on a per-trigger basis by passing a timeout in the manifest you submit.

Specific to 3, I'm happy to change it to read from an environment variable rather than a flag. Let me know

Comment (zroubalik) [reply adapter/main.go:112]: I like the environment variable approach as well.

**Comment by arschles** [comment]
@aman-bansal @zroubalik I commited a change for both reading the timeout from an env var and also putting the HTTP client into the ScalerConfig (instead of changing the signatures for each scaler constructor). 

One question that I still have is, in the Metrics API scaler, how should we configure the TLS transport? Currently, I have a TODO on it because if you configure it in the constructor, you'd be setting it on the global HTTP client. See https://github.com/kedacore/keda/pull/1251/files#diff-c832e36c124014f31496fdbc15f77397f48b73855a7b7441fd95525ed6112f54R82

**Comment by zroubalik** [comment]
@arschles I think that we should expose the env variable in Deployment yaml file (bc Helm chart is not the only installation method).

Sorry I missed the TLS transport question, do you have any particular solution in mind?

**Comment by aman-bansal** [comment]
I don't think we can do this. TLS config is something which is scaler specific. It needs to be manager per scaler basis. One single http.Client can't handle multiple TLS configs. 
How about this for an approach:
Just like pkg/scaling/resolver/scale_resolvers.go:110#resolveEnv resolves env variables for a container and add it to the scaler config, entertain one more case where env variable is read from keda container. It will become part of scaler config. Now for every scaler you have this information available. Not just initialise each client per scaler.

**Comment by arschles** [comment]
@aman-bansal the metrics API scaler was the only one I saw that needs to specify its own TLS config (see my comment in that file) - would it make sense to make that a special case for now, instead of creating a client per scaler? I'm happy to do either, but the connection pooling across scalers could be beneficial. I'm not familiar with common workloads for KEDA, so I'll let you all decide whether that is truly beneficial

**Comment by aman-bansal** [comment]
currently we have tls support in metric api scaler and kafka scaler. There is one open PR for rabbit mq scaler.

**Comment by arschles** [comment]
@aman-bansal got it. I'll change this to create one client per scaler like you said then.

**Comment by arschles** [comment]
@aman-bansal I've updated to create a new HTTP client for each scaler, instead of sharing one across them all. PTAL :smile:

**Comment by arschles** [comment]
Corresponding helm chart update: https://github.com/kedacore/charts/pull/90

**Comment by arschles** [comment]
Just fixed all the golang-ci checks. Looks like tests are broken on a context deadline exceeded error, but I cannot repro locally. am I not setting something up locally to be able to talk to rabbit?

File: pkg/scalers/http_clients.go
Comment (zroubalik) [inline pkg/scalers/http_clients.go:1]: Just a small nit, could you please move this file out of the `scalers` package, preferably somewhere under `pkg/util`?

File: pkg/scalers/metrics_api_scaler.go
Comment (zroubalik) [inline pkg/scalers/metrics_api_scaler.go:80]: We can remove this TODO now, right?

**Comment by zroubalik** [comment]
> Just fixed all the golang-ci checks. Looks like tests are broken on a context deadline exceeded error, but I cannot repro locally. am I not setting something up locally to be able to talk to rabbit?

That was just some intermittent failure caused by the environment, I've rescheduled the checks and it is passing now.

**Comment by zroubalik** [comment]
Could you please update Changelog as well?

**Comment by arschles** [comment]
@zroubalik I've been busy working on the HTTP scaling stuff but I haven't forgotten about this. I am going to be on vacation for the rest of the week and I'll get back to this early next.

File: pkg/scalers/http_clients.go
Comment (arschles) [reply pkg/scalers/http_clients.go]: done in https://github.com/kedacore/keda/pull/1251/commits/833043a665b27a8909e3667fe016481efd288d78

File: pkg/scalers/metrics_api_scaler.go
Comment (arschles) [reply pkg/scalers/metrics_api_scaler.go]: yup, done in https://github.com/kedacore/keda/pull/1251/commits/833043a665b27a8909e3667fe016481efd288d78

**Comment by arschles** [comment]
@zroubalik ok, just made all the changes you requested and resolved conflicts. PTAL

**Comment by arschles** [comment]
@zroubalik alright, I had time to give this my full attention and it should be fixed up now (I foolishly didn't rebase the first time 😦). PTAL

**Comment by zroubalik** [comment]
@arschles one more small rebase and I think we are good to merge! Thanks 👍 

We should document the `KEDA_HTTP_DEFAULT_TIMEOUT` env variable somewhere though, I am thinking about this section: https://keda.sh/docs/2.0/operate/     WDYT @tomkerkhove ?

**Comment by tomkerkhove** [comment]
> We should document the `KEDA_HTTP_DEFAULT_TIMEOUT` env variable somewhere though, I am thinking about this section: [keda.sh/docs/2.0/operate](https://keda.sh/docs/2.0/operate/) WDYT @tomkerkhove ?

Sounds like a good spot to me indeed!

**Comment by arschles** [comment]
Great, I'll fix this up today and add to the docs. Thanks @zroubalik @tomkerkhove 😄

**Comment by zroubalik** [comment]
Looking good! let's merge this once the docs are out :)

**Comment by arschles** [comment]
@zroubalik @tomkerkhove see https://github.com/kedacore/keda-docs/pull/324 for the docs PR

**Review by zroubalik — APPROVED** [review]
LGTM, thanks a lot @arschles for this!

**Pull request merged**
